### ISSUE TITLE
🐛 [Crash] puppetutils::setattAchment [commit 04cab83]

### ISSUE THREAD

**Issue by kaincenteno** [DerpyProjectGroup/topaz#2115]
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have searched existing [issues](https://github.com/LandSandBoat/server/issues) to see if the issue has already been opened, and I have checked the commit log to see if the issue has been resolved since my server was last updated
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)

## Branch affected by issue 

`base`

## Steps to reproduce

couldnt reach out to all the players but providing backtrace info bellow

## Expected behavior

no crash

```
(gdb) bt full
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
        set = {__val = {0, 9223372034707292159, 3277868454, 0, 3278190238, 0, 3211903488, 0, 1045220608, 0, 3278190238, 0, 8458093748099638637, 7593474027732954996, 3225419776, 0}}
        pid = <optimized out>
        tid = <optimized out>
        ret = <optimized out>
#1  0x00007fc439259859 in __GI_abort () at abort.c:79
        save_stage = 1
        act = {__sigaction_handler = {sa_handler = 0x7fc4394245c0 <_IO_2_1_stderr_>, sa_sigaction = 0x7fc4394245c0 <_IO_2_1_stderr_>}, sa_mask = {__val = {140480754219024, 3432, 140480749529048, 45, 1083757432, 1083757432, 
              140480750962112, 45, 45, 1, 140480754219024, 140480750945440, 140480749478897, 140480444669808, 1083757432, 3}}, sa_flags = 1095911088, sa_restorer = 0x4098d378}
        sigs = {__val = {32, 0 <repeats 15 times>}}
#2  0x00007fc439656911 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#3  0x00007fc43966238c in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#4  0x00007fc4396623f7 in std::terminate() () from /lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#5  0x000000000040e32b in __clang_call_terminate ()
No symbol table info available.
#6  0x00000000005f375c in sol::detail::static_trampoline<&sol::function_detail::upvalue_this_member_function<CLuaBaseEntity, void (CLuaBaseEntity::*)(unsigned short, short)>::real_call> (L=0x4098d378)
    at /server/ext/sol/include/sol/sol.hpp:8771
No locals.
#7  sol::detail::typed_static_trampoline<int (*)(lua_State*), &sol::function_detail::upvalue_this_member_function<CLuaBaseEntity, void (CLuaBaseEntity::*)(unsigned short, short)>::real_call> (L=0x4098d378)
    at /server/ext/sol/include/sol/sol.hpp:8893
No locals.
#8  sol::function_detail::upvalue_this_member_function<CLuaBaseEntity, void (CLuaBaseEntity::*)(unsigned short, short)>::call<false, false> (L=0x4098d378) at /server/ext/sol/include/sol/sol.hpp:19378
        nr = <optimized out>
#9  0x00007fc4397f9fb7 in ?? () from /lib/x86_64-linux-gnu/libluajit-5.1.so.2
No symbol table info available.
#10 0x00007fc4398484a4 in lua_pcall () from /lib/x86_64-linux-gnu/libluajit-5.1.so.2
No symbol table info available.
#11 0x00000000006b6be0 in sol::basic_protected_function<sol::stack_reference, true, sol::basic_reference<false> >::luacall<true> (this=0x7ffc71f37410, argcount=2, result_count_=-1, h=...)
    at /server/ext/sol/include/sol/sol.hpp:21088
No locals.
#12 sol::basic_protected_function<sol::stack_reference, true, sol::basic_reference<false> >::invoke<true>(sol::types<>, std::integer_sequence<unsigned long>, long, sol::detail::protected_handler<true, sol::basic_reference<false> >&) const (this=0x7ffc71f37410, n=2, h=...) at /server/ext/sol/include/sol/sol.hpp:21119
        stacksize = <optimized out>
        poststacksize = <optimized out>
        firstreturn = 1
        returncount = 0
        code = sol::call_status::ok
#13 sol::basic_protected_function<sol::stack_reference, true, sol::basic_reference<false> >::call<, CLuaBaseEntity, CLuaItem>(CLuaBaseEntity&&, CLuaItem&&) const (this=0x7ffc71f37410, args=..., args=...)
    at /server/ext/sol/include/sol/sol.hpp:21027
        h = {m_L = 0x4098d378, target = <optimized out>, stack_index = 1}
        pushcount = 2
#14 0x0000000000689c27 in sol::table_proxy<sol::basic_table_core<true, sol::basic_reference<false> >&, std::tuple<char const (&) [3], char const (&) [8], char const (&) [10], char const (&) [5], char const (&) [12], char const*&, char const (&) [10]> >::call<, CLuaBaseEntity, CLuaItem>(CLuaBaseEntity&&, CLuaItem&&) (this=<optimized out>, args=..., args=...) at /server/ext/sol/include/sol/sol.hpp:24897
        func = {<sol::basic_object<sol::stack_reference>> = {<sol::basic_object_base<sol::stack_reference>> = {<sol::stack_reference> = {<sol::stateless_stack_reference> = {m_index = 1}, 
                luastate = 0x4098d378}, <No data fields>}, <No data fields>}, static is_stack_handler_v = false, m_error_handler = {<sol::stateless_reference> = {ref = 151346}, luastate = 0x4098d378}}
        L = 0x4098d378
--Type <RET> for more, q to quit, c to continue without paging--
        idx = <optimized out>
#15 sol::table_proxy<sol::basic_table_core<true, sol::basic_reference<false> >&, std::tuple<char const (&) [3], char const (&) [8], char const (&) [10], char const (&) [5], char const (&) [12], char const*&, char const (&) [10]> >::operator()<CLuaBaseEntity, CLuaItem> (this=<optimized out>, args=..., args=...) at /server/ext/sol/include/sol/sol.hpp:24902
No locals.
#16 luautils::OnAttachmentUnequip (PEntity=<optimized out>, attachment=<optimized out>) at /server/src/map/lua/luautils.cpp:2370
        result = {<sol::proxy_base<sol::protected_function_result>> = {<sol::proxy_base_tag> = {<No data fields>}, <No data fields>}, L = 0x4098d378, index = 1, returncount = 0, popcount = 0, err = sol::call_status::ok}
        onUnequip = <optimized out>
        name = <optimized out>
#17 0x0000000000766987 in puppetutils::setAttachment (PChar=<optimized out>, slotId=<optimized out>, attachment=<optimized out>) at /server/src/map/utils/puppetutils.cpp:315
        PAttachment = 0x0
        oldAttachment = <optimized out>
#18 0x00000000007fefb0 in SmallPacket0x102 (PSession=<optimized out>, PChar=0x7fc4271f2990, data=...) at /server/src/map/packet_system.cpp:6944
        i = 22 '\026'
        attachment = <optimized out>
        job = <optimized out>
#19 0x0000000000805ad0 in std::_Function_handler<void (map_session_data_t*, CCharEntity*, CBasicPacket), void (*)(map_session_data_t*, CCharEntity*, CBasicPacket)>::_M_invoke(std::_Any_data const&, map_session_data_t*&&, CCharEntity*&&, CBasicPacket&&) (__functor=..., __args=..., __args=..., __args=...) at /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/std_function.h:300
No locals.
#20 0x00000000007ac90a in std::function<void (map_session_data_t*, CCharEntity*, CBasicPacket)>::operator()(map_session_data_t*, CCharEntity*, CBasicPacket) const (this=<optimized out>, __args=..., __args=..., __args=...)
    at /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/std_function.h:688
No locals.
#21 parse (buff=0x7fc42776f6a0 "y", buffsize=<optimized out>, from=<optimized out>, map_session_data=0x7fc4271b6010) at /server/src/map/map.cpp:746
        SmallPD_ptr = 0x7fc42776f6dc "\002Sy"
        PChar = 0x7fc4271f2990
        PacketData_End = 0x7fc42776f782 "\r_\372^\343\317\371;\222\252\307\313\247SX\216?\022E\b\277S\005\017\060F\260l\342\210\036W8uWB\230d\350\062\264\260\021\312\032\366\350\254\244}\030\344\256\220+\262\360x\375E4\377\220\t?\240\301\271\202Jx\343\377\247\215\037\030?\306\233\244\210\066v:\272_904\201\301\357\244w&\367\037d95\210\315\065S\267k\215\353o\016\256\370\212\272\233\266=\037\273\261S\001\300\202\002\250\060\262w"
        PacketData_Begin = <optimized out>
        SmallPD_Type = <optimized out>
        SmallPD_Size = <optimized out>
        SmallPD_Code = 121
        SmallPD_ptr = <optimized out>
        _msgStr = <optimized out>
        _msgStr = <optimized out>
        _msgStr = <optimized out>
        _msgStr = <optimized out>
        _msgStr = <optimized out>
#22 do_sockets (rfd=<optimized out>, next=...) at /server/src/map/map.cpp:512
        data = <optimized out>
        ipp = <optimized out>
        map_session_data = 0x7fc4271b6010
        ip = <optimized out>
        port = <optimized out>
        size = <optimized out>
        from = {sin_family = 2, sin_port = 19155, sin_addr = {s_addr = 824216088}, sin_zero = "\000\000\000\000\000\000\000"}
        fromlen = 16
        ret = <optimized out>
        timeout = {tv_sec = 0, tv_usec = 77455}
        ret = <optimized out>
#23 0x000000000041e11b in main (argc=1, argv=<optimized out>) at /server/src/common/kernel.cpp:278
--Type <RET> for more, q to quit, c to continue without paging--
        next = {__r = 0}
        rfd = {fds_bits = {8192, 0 <repeats 15 times>}}
```

**Comment by RAIST5150** [comment]
Calamity on xiweb discord (general bugs) recently noted it appears to consistently happen with light/dark based attachments.

**Comment by m241dan** [comment]
I believe it's because the regen and refresh attachments have fucking nils in the table. Those get passed to delMod, which expects CLuaBaseEntity + int, and not CLuaBaseEntity + NIL, and so sol freaks out (not sure why it doesn't freak out gracefully).

You'll notice on the "adding" of regen and refresh there are a whole load of complexity that is not represented in the removal process. The problem may be that  that maneuvers aren't actually passed to the removal call.

Perhaps someone who has some domain knowledge here could assist? @claywar 

Should be able to replicate by trying to remove a regen / refresh attachment.

**Comment by claywar** [comment]
Nope.
```
        -- Get base modifier value
        if modList[1] == xi.mod.REGEN then
            modValue = getRegenModValue(pet, attachmentName, maneuvers)
        elseif modList[1] == xi.mod.REFRESH then
            modValue = getRefreshModValue(pet, attachmentName, maneuvers)
        else
            modValue = modList[2][maneuvers + 1]
        end
```

**Comment by m241dan** [comment]
Yup.

```
xi.automaton.onAttachmentUnequip = function(pet, attachment)
    local modTable = attachmentModifiers[attachment:getName()]

    for k, modList in ipairs(modTable) do
        pet:delMod(modList[1], modList[2][1])
    end
end
```

**Comment by claywar** [comment]
You are correct that there is an issue here; however, it is unrelated to this crash and lua handles it gracefully.

**Comment by m241dan** [comment]
> You are correct that there is an issue here; however, it is unrelated to this crash and lua handles it gracefully.

You sure? Because when I stopped passing the nil, my server stopped crashing.

**Comment by claywar** [comment]
No crashes locally unless there are additional steps to the procedure.  Lua bails out of the unequipped attachment
![image](https://user-images.githubusercontent.com/81713309/177011063-b0441f9c-27d4-4209-bf0e-5a973c857020.png)

**Comment by m241dan** [comment]
Windows or Linux?

**Comment by kaincenteno** [comment]
@m241dan  a nil is crashing your server?  dunno if related but https://github.com/LandSandBoat/server/issues/2116 since those scripts dont exist so i assume it's also a nil.

**Comment by m241dan** [comment]
> @m241dan a nil is crashing your server? dunno if related but #2116 since those scripts dont exist so i assume it's also a nil.

Well, no longer passing the nil on removal stops the crash. I just hard set that mod to 0 on regen / refresh. Little hack for now. I could set something up and randomly pass nil on a range of functions and see if that crashes them. To rule it out.

I do know, when I booted with GDB, and ran the whole thing back. The entire stack never exits the SOL chain before crashing. So, it is something in SOL that is crashing.

Could be SOL is handling `nil` gracefully for Windows but not Linux.

**Comment by kaincenteno** [comment]
> > @m241dan a nil is crashing your server? dunno if related but #2116 since those scripts dont exist so i assume it's also a nil.
> 
> Well, no longer passing the nil on removal stops the crash. I just hard set that mod to 0 on regen / refresh. Little hack for now. I could set something up and randomly pass nil on a range of functions and see if that crashes them. To rule it out.
> 
> I do know, when I booted with GDB, and ran the whole thing back. The entire stack never exits the SOL chain before crashing. So, it is something in SOL that is crashing.
> 
> Could be SOL is handling `nil` gracefully for Windows but not Linux.

yeah i was thinking that same thing too but devs were unable to repro on their end. but always the crash logs are useful to them =)

Can you do me a favour and on your server ` !gotoid 17666576` and talk to the question mark target

**Comment by m241dan** [comment]
> > > @m241dan a nil is crashing your server? dunno if related but #2116 since those scripts dont exist so i assume it's also a nil.
> > 
> > 
> > Well, no longer passing the nil on removal stops the crash. I just hard set that mod to 0 on regen / refresh. Little hack for now. I could set something up and randomly pass nil on a range of functions and see if that crashes them. To rule it out.
> > I do know, when I booted with GDB, and ran the whole thing back. The entire stack never exits the SOL chain before crashing. So, it is something in SOL that is crashing.
> > Could be SOL is handling `nil` gracefully for Windows but not Linux.
> 
> yeah i was thinking that same thing too but devs were unable to repro on their end. but always the crash logs are useful to them =)
> 
> Can you do me a favour and on your server ` !gotoid 17666576` and talk to the question mark target

I will tomorrow, I just shut my whole dev setup down and am playing now.

**Comment by claywar** [comment]
Use the local functions to determine the value of regen or refresh, else you're opening yourself for some exploits with insane regen/refresh by unequipping/re-equipping attachments.

**Comment by m241dan** [comment]
> Use the local functions to determine the value of regen or refresh, else you're opening yourself for some exploits with insane regen/refresh by unequipping/re-equipping attachments.

Have to pass maneuvers through, and unequip doesn't have those. I haven't dug into what it would take to add that to the unequip. I just sort of figured the person with better domain knowledge (you) would be the best to fix this. But, I'll take a crack at some point. Not a priority for me.

**Comment by claywar** [comment]
Its on my list to fix.  Maneuver count _should_ always be 0 on removing an attachment, as we should be removing those effects on deactivate.  Needs confirmation, but I believe is the case.

**Comment by m241dan** [comment]
> !gotoid 17666576

Crashed.

Also whipped up this little beauty:

```
-----------------------------------
-- func: delmodnil
-- desc: Tests passing nil to delMod to see if it crashes linux system
-----------------------------------

require("scripts/globals/status")

cmdprops =
{
    permission = 1,
    parameters = ""
}


function onTrigger(player)
    player:delMod(xi.mod.REGEN, nil)
end
```

It also crashes. So, SOL hates Linux?

**Comment by WinterSolstice8** [comment]
```[07/03/22 09:24:34:906][map][error] cmdhandler::call: (delmodnil) error: stack index 3, expected number, received nil: not a numeric type (bad argument into 'void(short unsigned int, short int)')
stack traceback:
        [C]: in function 'delMod'
        ./scripts/commands/delmodnil.lua:16: in function <./scripts/commands/delmodnil.lua:15> (call:321)
```

Works on my machine.

post the output of these commands:

`uname -a`

```
 uname -a
Linux vps 5.15.0-1011-oracle #15-Ubuntu SMP Fri Jun 10 12:21:49 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux
```

`cat /etc/lsb-release` (some distros wont have this, so use /etc/issue instead if that's the case)
 
```
cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04 LTS"
```

and finally

cd to xi_map folder

`/usr/bin/ldd ./xi_map`

```
/usr/bin/ldd ./xi_map
        linux-vdso.so.1 (0x0000ffff86f84000)
        libzmq.so.5 => /lib/aarch64-linux-gnu/libzmq.so.5 (0x0000ffff86970000)
        libluajit-5.1.so.2 => /lib/aarch64-linux-gnu/libluajit-5.1.so.2 (0x0000ffff868e0000)
        libmariadb.so.3 => /lib/aarch64-linux-gnu/libmariadb.so.3 (0x0000ffff86880000)
        libstdc++.so.6 => /lib/aarch64-linux-gnu/libstdc++.so.6 (0x0000ffff86650000)
        libm.so.6 => /lib/aarch64-linux-gnu/libm.so.6 (0x0000ffff865b0000)
        libgcc_s.so.1 => /lib/aarch64-linux-gnu/libgcc_s.so.1 (0x0000ffff86580000)
        libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000ffff863d0000)
        /lib/ld-linux-aarch64.so.1 (0x0000ffff86f4b000)
        libbsd.so.0 => /lib/aarch64-linux-gnu/libbsd.so.0 (0x0000ffff863a0000)
        libsodium.so.23 => /lib/aarch64-linux-gnu/libsodium.so.23 (0x0000ffff86350000)
        libpgm-5.3.so.0 => /lib/aarch64-linux-gnu/libpgm-5.3.so.0 (0x0000ffff862f0000)
        libnorm.so.1 => /lib/aarch64-linux-gnu/libnorm.so.1 (0x0000ffff861b0000)
        libgssapi_krb5.so.2 => /lib/aarch64-linux-gnu/libgssapi_krb5.so.2 (0x0000ffff86150000)
        libdl.so.2 => /lib/aarch64-linux-gnu/libdl.so.2 (0x0000ffff86130000)
        libssl.so.3 => /lib/aarch64-linux-gnu/libssl.so.3 (0x0000ffff86080000)
        libcrypto.so.3 => /lib/aarch64-linux-gnu/libcrypto.so.3 (0x0000ffff85c90000)
        libz.so.1 => /lib/aarch64-linux-gnu/libz.so.1 (0x0000ffff85c60000)
        libmd.so.0 => /lib/aarch64-linux-gnu/libmd.so.0 (0x0000ffff85c40000)
        libpthread.so.0 => /lib/aarch64-linux-gnu/libpthread.so.0 (0x0000ffff85c20000)
        libkrb5.so.3 => /lib/aarch64-linux-gnu/libkrb5.so.3 (0x0000ffff85b40000)
        libk5crypto.so.3 => /lib/aarch64-linux-gnu/libk5crypto.so.3 (0x0000ffff85b00000)
        libcom_err.so.2 => /lib/aarch64-linux-gnu/libcom_err.so.2 (0x0000ffff85ae0000)
        libkrb5support.so.0 => /lib/aarch64-linux-gnu/libkrb5support.so.0 (0x0000ffff85ac0000)
        libkeyutils.so.1 => /lib/aarch64-linux-gnu/libkeyutils.so.1 (0x0000ffff85aa0000)
        libresolv.so.2 => /lib/aarch64-linux-gnu/libresolv.so.2 (0x0000ffff85a70000)
```

**Comment by WinterSolstice8** [comment]
Works in debug, release with debug symbols, release mode for me. So no variance there.

**Comment by m241dan** [comment]
Hmmm, we updated sol 8 days ago. I have not done that yet, maybe that's part of the problem?

**Comment by zach2good** [comment]
Why would you update sol? It's bundled on all platforms

**Comment by m241dan** [comment]
> 

I didn't, you did? That was a "you" == "we". 

https://github.com/LandSandBoat/server/commit/2073cf5e3564308a70ef98ef6b14cf9562d74355

**Comment by zach2good** [comment]
Accidental close, but  "you" != "we".

**Comment by m241dan** [comment]
"we" as in the project... but you specifically did it.

**Comment by WinterSolstice8** [comment]
Post the output of the requested commands.

**Comment by m241dan** [comment]
> Post the output of the requested commands.

Yes sir! lol

uname -a
```
Linux DESKTOP-014N7V2 5.10.16.3-microsoft-standard-WSL2 #1 SMP Fri Apr 2 22:23:49 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

cat /etc/lsb-release
```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.4 LTS"
```

/usr/bin/ldd ./xi_map
```
        linux-vdso.so.1 (0x00007ffc3cb7a000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007ff2ddbe2000)
        libzmq.so.5 => /lib/x86_64-linux-gnu/libzmq.so.5 (0x00007ff2ddb3c000)
        libluajit-5.1.so.2 => /lib/x86_64-linux-gnu/libluajit-5.1.so.2 (0x00007ff2ddac1000)
        libmariadb.so.3 => /lib/x86_64-linux-gnu/libmariadb.so.3 (0x00007ff2dda6a000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007ff2dd888000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007ff2dd739000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007ff2dd71c000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007ff2dd6f9000)
        /lib64/ld-linux-x86-64.so.2 (0x00007ff2dec35000)
        libsodium.so.23 => /lib/x86_64-linux-gnu/libsodium.so.23 (0x00007ff2dd6a1000)
        libpgm-5.2.so.0 => /lib/x86_64-linux-gnu/libpgm-5.2.so.0 (0x00007ff2dd652000)
        libnorm.so.1 => /lib/x86_64-linux-gnu/libnorm.so.1 (0x00007ff2dd4f7000)
        libgssapi_krb5.so.2 => /lib/x86_64-linux-gnu/libgssapi_krb5.so.2 (0x00007ff2dd4aa000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007ff2dd4a2000)
        libgnutls.so.30 => /lib/x86_64-linux-gnu/libgnutls.so.30 (0x00007ff2dd2cc000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007ff2dd2b0000)
        libkrb5.so.3 => /lib/x86_64-linux-gnu/libkrb5.so.3 (0x00007ff2dd1d3000)
        libk5crypto.so.3 => /lib/x86_64-linux-gnu/libk5crypto.so.3 (0x00007ff2dd1a2000)
        libcom_err.so.2 => /lib/x86_64-linux-gnu/libcom_err.so.2 (0x00007ff2dd199000)
        libkrb5support.so.0 => /lib/x86_64-linux-gnu/libkrb5support.so.0 (0x00007ff2dd18a000)
        libp11-kit.so.0 => /lib/x86_64-linux-gnu/libp11-kit.so.0 (0x00007ff2dd054000)
        libidn2.so.0 => /lib/x86_64-linux-gnu/libidn2.so.0 (0x00007ff2dd033000)
        libunistring.so.2 => /lib/x86_64-linux-gnu/libunistring.so.2 (0x00007ff2dceb1000)
        libtasn1.so.6 => /lib/x86_64-linux-gnu/libtasn1.so.6 (0x00007ff2dce9b000)
        libnettle.so.7 => /lib/x86_64-linux-gnu/libnettle.so.7 (0x00007ff2dce5f000)
        libhogweed.so.5 => /lib/x86_64-linux-gnu/libhogweed.so.5 (0x00007ff2dce28000)
        libgmp.so.10 => /lib/x86_64-linux-gnu/libgmp.so.10 (0x00007ff2dcda4000)
        libkeyutils.so.1 => /lib/x86_64-linux-gnu/libkeyutils.so.1 (0x00007ff2dcd9d000)
        libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007ff2dcd81000)
        libffi.so.7 => /lib/x86_64-linux-gnu/libffi.so.7 (0x00007ff2dcd73000)
```

**Comment by m241dan** [comment]
Crashes in my docker container too, but I'm feeling too lazy to start the run-time up but here's the dockerfile.

In general, the only real difference I'm seeing is Ubuntu 20.04 and I don't have the most recent sol.hpps that Z2G updated 8 days ago.

```
FROM ubuntu:20.04 as dev-env

ENV DEBIAN_FRONTEND=noninteractive

RUN apt update && apt install -y --no-install-recommends \
    build-essential \
    ca-certificates \
    git \
    g++-9 \
    cmake \
    make \
    gdb \
    libluajit-5.1-dev \
    libzmq3-dev \
    libssl-dev  \
    zlib1g-dev  \
    mariadb-server  \
    libmariadb-dev \
    && apt clean

FROM dev-env as build

# Do building stuff

ADD ./cmake /app/cmake
ADD ./ext /app/ext
ADD ./src /app/src
ADD ./modules /app/modules
ADD ./CMakeLists.txt /app

WORKDIR /app

# set on commandline with "--build-arg num_jobs=<your number>" can be $(nproc)
ARG num_jobs=1

# build and clean up
RUN mkdir build && cd build && cmake .. && cmake --build . -- -j $num_jobs && cd .. && rm -r build

FROM ubuntu:20.04 as run-time

ENV DEBIAN_FRONTEND=noninteractive

RUN apt update && apt install -y --no-install-recommends \
    libmariadb3 \
    libzmq5 \
    libluajit-5.1-2 \
    mysql-client \
    && apt clean


COPY --from=build /app/xi_connect /
COPY --from=build /app/xi_map /
COPY --from=build /app/xi_search /

ADD ./conf /conf
ADD ./scripts /scripts
ADD ./modules /modules
ADD ./navmeshes /navmeshes
ADD ./tools/wait_for_silly_sql.sh /wait_for_silly_sql.sh
ADD ./compress.dat /compress.dat
ADD ./decompress.dat /decompress.dat
```

**Comment by WinterSolstice8** [comment]
Hmmm. I could not get 20.04 to build LSB (at least on ARM, I didn't test x84). Did you do anything special to make it work? It works out of the box on 22.04 following the commands to get prereqs.

It may be worth upgrading to 22.04, though

https://www.linuxtechi.com/upgrade-ubuntu-20-04-to-ubuntu-22-04/

**Comment by m241dan** [comment]
> Hmmm. I could not get 20.04 to build LSB (at least on ARM, I didn't test x84). Did you do anything special to make it work? It works out of the box on 22.04 following the commands to get prereqs.
> 
> It may be worth upgrading to 22.04, though
> 
> https://www.linuxtechi.com/upgrade-ubuntu-20-04-to-ubuntu-22-04/

Ya, certainly worth considering. Gonna rebase what I'm missing first and go from there. Shouldn't be going on about things crashing without being up to date with the most recent stuff anyway. But technically, this isn't my issue, just sharing my experience.

**Comment by WinterSolstice8** [comment]
Differences I see in ldd:

you are using gnutls and not openssl ( I doubt this makes a difference, but it is one), which causes several dependency shifts
libpgm 5.2 -> 5.3, also probably unrelated.

WSL is _supposed_ to be a full kernel replacement for Linux but I would not be shocked if at runtime some things might work differently. All in all, nothing concrete but interesting. The names of the .so don't correlate to bugfixes, though. It may be possible that newer packages roll in a ton of bugfixes we can't see without running exhaustive commands

**Comment by m241dan** [comment]
> Differences I see in ldd:
> 
> you are using gnutls and not openssl ( I doubt this makes a difference, but it is one), which causes several dependency shifts libpgm 5.2 -> 5.3, also probably unrelated.
> 
> WSL is _supposed_ to be a full kernel replacement for Linux but I would not be shocked if at runtime some things might work differently. All in all, nothing concrete but interesting. The names of the .so don't correlate to bugfixes, though. It may be possible that newer packages roll in a ton of bugfixes we can't see without running exhaustive commands

Yah, I mean, we'll figure it out eventually. Ideally , nil's shouldn't be being passed at all ever, but ya... ideal world, etc.

**Comment by WinterSolstice8** [comment]
alright so

x86 WSL : crash
x86 bare metal : crash
aarch64 (not bare metal, some kind of hypervisor): no crash

will check bare metal aarch64 with my raspi4 later...

**Comment by kaincenteno** [comment]
> alright so
> 
> x86 WSL : crash x86 bare metal : crash aarch64 (not bare metal, some kind of hypervisor): no crash
> 
> will check bare metal aarch64 with my raspi4 later...

Winter did you also test https://github.com/LandSandBoat/server/issues/2116 on aarch64? if so can you retest on x86?

**Comment by WinterSolstice8** [comment]
No I will not retest.

**Comment by WinterSolstice8** [comment]
works on bare metal aarch64. will get back to everyone on this later

**Comment by WinterSolstice8** [comment]
I now understand the root cause of the crash. It's an external library. Need some internal discussion on how to move forward.

**Comment by WinterSolstice8** [comment]
probably fixed by #2354
